### ISSUE TITLE
Rollout can sometimes run indefinitely when new pods are in a crash loop

### ISSUE THREAD

**Issue by andrii-korotkov-verkada** [argoproj/argo-rollouts#3765]
Checklist:

* [x] I've included steps to reproduce the bug.
* [x] I've included the version of argo rollouts.

**Describe the bug**

There's an occasional issue with ArgoCD sync can run for 1+ day waiting for a healthy state of a rollout, while the new version pods are crash looping. The rollout is stuck at the first canary step and neither progresses to the next step nor rolls back. The number of ready replicas keeps going up and down by a few, which might be confusing the rollouts controller that some progress is actually happening.
Note: that's a different issue from previously reported rollouts stuck due to "object modified", there's no indication in the logs that that's happening.

**To Reproduce**

Create a deployment with pods that can run okay. Create a rollout with 100 replicas and a first canary step with 10% of pods to be updated first, no analysis run. Put some analysis run for the 2nd step and beyond, but not sure it's necessary. Get a new pod version that should crash on a startup and try to release it. May need to repeat a few times.

**Expected behavior**

The rollout should get automatically aborted with pods rolling back to a previous version, app sync fails and app enters a degraded state.

**Screenshots**

**Version**

v1.7.1

**Logs**

From oldest to newest (partial logs)
```
* Enqueueing parent of default/<new-replica-set-name>: Rollout default/<rollout-name>
* Patched: {"status":{"availableReplicas":92,"conditions":[{"lastTransitionTime":"2024-08-02T00:25:25Z","lastUpdateTime":"2024-08-02T00:25:25Z","message":"Rollout is paused","reason":"RolloutPaused","status":"False","type":"Paused"},{"lastTransitionTime":"2024-08-02T23:50:30Z","lastUpdateTime":"2024-08-02T23:50:30Z","message":"Rollout is not healthy","reason":"RolloutHealthy","status":"False","type":"Healthy"},{"lastTransitionTime":"2024-08-02T23:50:30Z","lastUpdateTime":"2024-08-02T23:50:30Z","message":"Rollout does not have minimum availability","reason":"AvailableReason","status":"False","type":"Available"},{"lastTransitionTime":"2024-08-02T23:53:21Z","lastUpdateTime":"2024-08-02T23:53:21Z","message":"RolloutCompleted","reason":"RolloutCompleted","status":"False","type":"Completed"},{"lastTransitionTime":"2024-08-02T17:21:27Z","lastUpdateTime":"2024-08-03T01:30:07Z","message":"ReplicaSet \"<new-replica-set-name>\" is progressing.","reason":"ReplicaSetUpdated","status":"True","type":"Progressing"}],"readyReplicas":92}}
* Enqueueing parent of default/<new-replica-set-name>: Rollout default/<rollout-name>
* Enqueueing parent of default/<new-replica-set-name>: Rollout default/<rollout-name>
* Enqueueing parent of default/<new-replica-set-name>: Rollout default/<rollout-name>
* Patched: {"status":{"availableReplicas":91,"conditions":[{"lastTransitionTime":"2024-08-02T00:25:25Z","lastUpdateTime":"2024-08-02T00:25:25Z","message":"Rollout is paused","reason":"RolloutPaused","status":"False","type":"Paused"},{"lastTransitionTime":"2024-08-02T23:50:30Z","lastUpdateTime":"2024-08-02T23:50:30Z","message":"Rollout is not healthy","reason":"RolloutHealthy","status":"False","type":"Healthy"},{"lastTransitionTime":"2024-08-02T23:50:30Z","lastUpdateTime":"2024-08-02T23:50:30Z","message":"Rollout does not have minimum availability","reason":"AvailableReason","status":"False","type":"Available"},{"lastTransitionTime":"2024-08-02T23:53:21Z","lastUpdateTime":"2024-08-02T23:53:21Z","message":"RolloutCompleted","reason":"RolloutCompleted","status":"False","type":"Completed"},{"lastTransitionTime":"2024-08-02T17:21:27Z","lastUpdateTime":"2024-08-03T01:31:21Z","message":"ReplicaSet \"<new-replica-set-name>\" is progressing.","reason":"ReplicaSetUpdated","status":"True","type":"Progressing"}],"readyReplicas":91}}
* Enqueueing parent of default/<new-replica-set-name>: Rollout default/<rollout-name>
* Started syncing Analysis at (2024-08-03 01:32:10.919158096 +0000 UTC m=+220587.756264716)
* No status changes. Skipping patch
* Started syncing Analysis at (2024-08-03 01:32:10.919769572 +0000 UTC m=+220587.756876202)
* No status changes. Skipping patch
* Reconciliation completed
* Reconciliation completed
```
Analysis-related logs were there for multiple versions, including a new version's analysis run.
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by andrii-korotkov-verkada** [comment]
I also find logs like `"message":"Rollout is not healthy","reason":"RolloutHealthy"` confusing, since they seem to send different signals.

**Comment by andrii-korotkov-verkada** [comment]
For now, I've set the following on Rollouts:
```
  progressDeadlineAbort: true
  progressDeadlineSeconds: 900
```

**Comment by monali-reddy** [comment]
@andrii-korotkov-verkada We are facing the same issue, the rollout stays in progressing state indefinitely and does not obey the  progressDeadlineSeconds value. Did anything help you get it working?

**Comment by rave2101** [comment]
@monali-reddy facing the same issue. We have

```
progressDeadlineAbort: true
progressDeadlineSeconds: 600
minReadySeconds: 60
```

@andrii-korotkov-verkada wanted to know if there is anything that you did from your end to fix this issue? Thanks!

**Comment by andrii-korotkov-verkada** [comment]
Which argo rollouts version is it? I think it's been okay with progress deadline for at least some 1.8 versions.

**Comment by rave2101** [comment]
We have argo rollouts version `1.8.3`

**Comment by rave2101** [comment]
I have opened a similar thread [here](https://github.com/argoproj/argo-rollouts/discussions/4449) @andrii-korotkov-verkada , let me know if I'm missing something in the rollout manifest. If the pod struggles to come up or stuck, then the CPU usage for the rollout goes high and we're getting stuck.

**Comment by RunlinXiao** [comment]
I'm also facing the same issue where the rollout get stuck the whole time and argo CPU went up to 100% , and I think there is an easier way to reproduce this: set a non-exist secret for your container, e.g.
```
        - name: test-secret
          valueFrom:
            secretKeyRef:
              key: empty
              name: not-exist
```

**Comment by monali-reddy** [comment]
@andrii-korotkov-verkada I tried with argo rollouts version 1.7.2 and 1.8.3 and it's the same for both. Would I be missing something?

**Comment by rave2101** [comment]
> I'm also facing the same issue where the rollout get stuck the whole time and argo CPU went up to 100% , and I think there is an easier way to reproduce this: set a non-exist secret for your container, e.g.
> 
> ```
>         - name: test-secret
>           valueFrom:
>             secretKeyRef:
>               key: empty
>               name: not-exist
> ```

Yes, I’ve already verified that and it’s true. The issue occurs when a non-existent secret is referenced in the pod template. As mentioned in the comment above, I’m seeing the same behavior:
Even though the secret doesn’t exist, Argo is currently not detecting it.
The rollout still considers the pod as Progressing instead of triggering a progressDeadlineAbort.
In other words, the missing secret is present in the pod template, but Argo does not recognize it as an error condition and the rollout keeps running and then CPU usage also increases.

cc : @kostis-codefresh

**Comment by monali-reddy** [comment]
Update - Verified and this works with argo rollouts version 1.8.3.

**Comment by nurus** [comment]
We're also experiencing this issue running `v1.8.3` with the following options:
```
progressDeadlineAbort: true
progressDeadlineSeconds: 360
```

When the pod is crashlopping on startup the status `phase: Progressing` never changes. It should respect the options above and detect it must abort the build and rollback. Some more details about our setup, we have a 15 minute pause after the first replica is up to do the canary analysis.

**Comment by bmlis** [comment]
It seems to be working through CLI and UI, while abort using `progressDeadlineX` goes into infinite loop messing up with CPU and logs.

What I think is happening is that CLI/UI path directly updates `status.abort` on CRD which is then picked by controller and properly handled by `rolloutContext.checkPausedConditions`.

We use blue/green, and in that path https://github.com/argoproj/argo-rollouts/blob/49fa1516cf71672b69e265267da4e1d16e1fe114/rollout/service.go#L94 if new RS is not saturated, it will not switch to old RS if I understand correctly. I'm also not sure whether the fact that pauseContext is ignored here is intended (I see that IsAborted is checked for canary for example).

Because abort is not handled properly, https://github.com/argoproj/argo-rollouts/blob/49fa1516cf71672b69e265267da4e1d16e1fe114/rollout/sync.go#L837-L846 will trigger rollout indefinitely and very often after `progressDeadlineSeconds` passes.

This is my limited understanding after debugging this for a bit on a group call because adding `progressDeadlineX` configuration actually cause an incident for us. I'm probably missing a lot of crucial context here, but maybe it's a good starting point for someone with more knowledge or time to look into this?

The issue occurred on 1.8.3, so I think it's safe to assume that it is still present.
